### PR TITLE
Resolve deps using DependencyManager found in PDE

### PR DESCRIPTION
**Pull request by haubi**
Seen with Eclipse 2023-06, the launch configurations created by LcDSL lacks some required bundles, but the "Select Required" button in PDE launch config editor does resolve them.
The idea now is to utilize the PDE internal DependencyResolver, which drives the "Select Required" button.

Unfortunately, PDE DependencyResolver omits unresolved bundles, which we need to retain when exporting products (towards BDeploy) based on launch configurations, for allowing them to run on different platforms than the export was performed on.

Also, PDE DependencyResolver does not allow to omit arbitrary bundles, required for the 'ignore' LcDSL keyword.

Hence, at first I copy in PDE DependencyResolver without further changes.
Then, adapt the local DependencyResolver for our needs, in a dedicated commit which potentially can be forwarded towards PDE upstream.

Beyond that, also utilize the PDE "Select Required" implementation to define start levels for special bundles.

And finally, remove everything needless from local DependencyResolver, to reduce imports from PDE internal as much as possible.

For completion, this feels worth an LcDSL revbump.

### CODE REVIEW

File: com.wamas.ide.launching/src/com/wamas/ide/launching/generator/DependencyResolver.xtend [hunk 1]
```diff
@@ -265,48 +266,13 @@ class DependencyResolver {
...
-			result.add(dep.getExporter());
-		}
-
-		return result;
+		val Set<BundleFilter> filters = newHashSet(BundleFilter.Factory.getBundleIsIgnoredFilter(toIgnore))
+		if (!addAll) filters.add(BundleFilter.Factory.bundleIsResolvedFilter)
+		val result = DependencyManager.findRequirementsClosure(preselected.toList, BundleFilter.Factory.and(filters),
+			DependencyManager.Options.INCLUDE_OPTIONAL_DEPENDENCIES, DependencyManager.Options.INCLUDE_NON_TEST_FRAGMENTS
+		)
+		toIgnore.forEach[result.remove(it)]
```
Comment (haubi) [inline com.wamas.ide.launching/src/com/wamas/ide/launching/generator/DependencyResolver.xtend:287]: Obsolete due to BundleFilter.Factory.getBundleIsIgnoredFilter(toIgnore) above

File: com.wamas.ide.launching/src/com/wamas/ide/launching/generator/StandaloneLaunchConfigGenerator.xtend [hunk 2]
```diff
@@ -45,16 +45,6 @@ class StandaloneLaunchConfigGenerator {
 	@Inject
 	extension IQualifiedNameProvider qnp
 
-	// also see org.eclipse.pde.internal.ui.editor.product.PluginConfigurationSection.getBundlesWithStartLevels()
-	// for a list of known default start-levels which should be present.
-	val knownStartLevels = newHashMap(
```
Comment (jukzi) [inline com.wamas.ide.launching/src/com/wamas/ide/launching/generator/StandaloneLaunchConfigGenerator.xtend:50]: why is that logic remove?

File: com.wamas.ide.launching/src/com/wamas/ide/launching/generator/BundleFilter.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+        public static BundleFilter getBundleIsResolvedFilter() {
+            return BundleDescription::isResolved;
+        }
+
+        public static BundleFilter getBundleIsNotIgnoredFilter(Collection<BundleDescription> toIgnore) {
+            return (b) -> !toIgnore.contains(b);
+        }
+
+        public static BundleFilter and(Iterable<? extends Predicate<BundleDescription>> predicates) {
```
Comment (jukzi) [inline com.wamas.ide.launching/src/com/wamas/ide/launching/generator/BundleFilter.java:23]: better use java.util.function.Predicate.and(Predicate<? super T>)

File: com.wamas.ide.launching/src/com/wamas/ide/launching/generator/BundleFilter.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.Collection;
+import java.util.function.Predicate;
+
+import org.eclipse.osgi.service.resolver.BundleDescription;
+
+public interface BundleFilter extends Predicate<BundleDescription> {
+
+    public static class Factory {
+
+        public static BundleFilter getBundleIsResolvedFilter() {
```
Comment (jukzi) [inline com.wamas.ide.launching/src/com/wamas/ide/launching/generator/BundleFilter.java:15]: not used?

File: com.wamas.ide.launching/src/com/wamas/ide/launching/generator/BundleFilter.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+package com.wamas.ide.launching.generator;
+
+import java.util.Collection;
+import java.util.function.Predicate;
+
+import org.eclipse.osgi.service.resolver.BundleDescription;
+
+public interface BundleFilter extends Predicate<BundleDescription> {
+
+    public static class Factory {
```
Comment (jukzi) [inline com.wamas.ide.launching/src/com/wamas/ide/launching/generator/BundleFilter.java:13]: the whole Factory feels like overkill, as everything here can be archived with java.util.function

File: com.wamas.ide.launching/src/com/wamas/ide/launching/generator/DependencyManager.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+import org.osgi.framework.wiring.BundleWiring;
+
+/**
+ * Utility class to return bundle description collections for a variety of
+ * dependency scenarios
+ *
+ * @noextend This class is not intended to be subclassed by clients.
+ * @noinstantiate This class is not intended to be instantiated by clients.
+ */
+public class DependencyManager {
```
Comment (jukzi) [inline com.wamas.ide.launching/src/com/wamas/ide/launching/generator/DependencyManager.java:44]: is this the copy & paste code? then at add the source as comment

**Review by jukzi — COMMENTED** [review]
Why are unresolved bundles needed?

**Comment by mduft** [comment]
There were good reasons for us back then to *not* use any of the PDE code. Mainly due to it sometimes being totally wrong :D If you change this, I suggest you do some thorough tests in a large workspace with many launch configurations, potentially also with historic versions of it (different target platform versions) to assure that things are working correctly. Also note that the dependency calculation is not only used to launch applications, but also by TEA to determine the content of binary builds of the same applications. In the past we had the situation (sometimes) that launching in the workspace worked, but binary was broken and the other way round.

TL;DR: please test this change even better than most others :D

File: com.wamas.ide.launching/src/com/wamas/ide/launching/generator/StandaloneLaunchConfigGenerator.xtend (hunk 2 above)
Comment (haubi) [reply com.wamas.ide.launching/src/com/wamas/ide/launching/generator/StandaloneLaunchConfigGenerator.xtend:50]: Because the entries as returned by DependencyResolver.resolveAndExpand() contain proper startlevels per this commit. But then, the workspace may contain such a special bundle, be it for debugging purposes only.
Moving the logic back here, leaving DependencyResolver alone.

File: com.wamas.ide.launching/src/com/wamas/ide/launching/generator/BundleFilter.java (hunk 3 above)
Comment (haubi) [reply com.wamas.ide.launching/src/com/wamas/ide/launching/generator/BundleFilter.java]: Yeah - failed to find the proper Xtend syntax in the first attempt. Now I can drop BundleFilter again.

File: com.wamas.ide.launching/src/com/wamas/ide/launching/generator/DependencyManager.java (hunk 6 above)
Comment (haubi) [reply com.wamas.ide.launching/src/com/wamas/ide/launching/generator/DependencyManager.java:46]: Ack

File: com.wamas.ide.launching/src/com/wamas/ide/launching/generator/StandaloneLaunchConfigGenerator.xtend (hunk 2 above)
Comment (haubi) [reply com.wamas.ide.launching/src/com/wamas/ide/launching/generator/StandaloneLaunchConfigGenerator.xtend:50]: It turns out that PDE's internal list of well-known bundles with startlevels is eventually incomplete, or at least eventually outdated.
Actually, it's a recommendation only, see also https://github.com/eclipse-pde/eclipse.pde/discussions/754.

Comment (haubi) [reply com.wamas.ide.launching/src/com/wamas/ide/launching/generator/StandaloneLaunchConfigGenerator.xtend:50]: Preserving that list locally, just updating based on what is found in Eclipse 2023-06.
